src/metrics.rs

```diff
@@ -0,0 +1,250 @@
+/// and transmitted periodically in batches
+pub fn submit(point: Point) {
+    let agent_mutex = get_singleton_agent();
+    let agent = agent_mutex.lock().unwrap();
+    agent.submit(point);
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::sync::atomic::{AtomicUsize, Ordering};
+
+    struct MockMetricsWriter {
+        points_written: AtomicUsize,
+    }
+    impl MockMetricsWriter {
+        fn new() -> MockMetricsWriter {
+            MockMetricsWriter {
+                points_written: AtomicUsize::new(0),
+            }
+        }
+
+        fn assert_points_written(&self, n: usize) {
+            assert_eq!(n, self.points_written.load(Ordering::Relaxed));
+        }
+    }
+
+    impl MetricsWriter for MockMetricsWriter {
+        fn write(&self, points: &mut Vec<Point>) {
+            assert!(!points.is_empty());
+
+            self.points_written
+                .fetch_add(points.len(), Ordering::SeqCst);
+
+            println!(
+                "Writing {} points ({} total)",
+                points.len(),
+                self.points_written.load(Ordering::SeqCst)
+            );
+            points.clear();
+        }
+    }
+
+    #[test]
+    fn test_submit() {
+        let writer = Arc::new(MockMetricsWriter::new());
+        let agent = MetricsAgent::new_with_args(writer.clone(), None);
+
+        for i in 0..42 {
+            let point = format!("measurement {}", i);
+            agent.submit(point);
+        }
+
+        agent.flush();
+        writer.assert_points_written(42);
+    }
+
+    #[test]
+    fn test_submit_with_delay() {
+        let writer = Arc::new(MockMetricsWriter::new());
+        let agent = MetricsAgent::new_with_args(writer.clone(), Some(Duration::from_millis(100)));
+
+        agent.submit("point 1".to_string());
+        thread::sleep(Duration::from_secs(2));
+        writer.assert_points_written(1);
+    }
+
+    #[test]
+    fn test_multithread_submit() {
+        let writer = Arc::new(MockMetricsWriter::new());
+        let agent = Arc::new(Mutex::new(MetricsAgent::new_with_args(
+            writer.clone(),
+            None,
+        )));
+
+        //
+        // Submit measurements from different threads
+        //
+        let mut threads = Vec::new();
+        for i in 0..42 {
+            let point = format!("measurement {}", i);
+            let agent = Arc::clone(&agent);
+            threads.push(thread::spawn(move || {
+                agent.lock().unwrap().submit(point);
+            }));
+        }
+
+        for thread in threads {
+            thread.join().unwrap();
+        }
+
+        agent.lock().unwrap().flush();
+        writer.assert_points_written(42);
+    }
+
+    #[test]
+    fn test_flush_before_drop() {
+        let writer = Arc::new(MockMetricsWriter::new());
+        {
+            let agent =
+                MetricsAgent::new_with_args(writer.clone(), Some(Duration::from_secs(9999999)));
+            agent.submit("point 1".to_string());
+        }
```

Comment:
Nice test. Think there's any way to ensure the singleton is cleaned up as well?

Reply:
Oh probably, I burnt like 3 hours yesterday on trying to do that before I decided I was out of my depth in Rust.  Happy to accept hints, otherwise I'll probably revisit this in a couple weeks when I have more of a clue!